packages/mui-material/src/Tabs/Tabs.js

```diff
@@ -614,29 +587,65 @@ const Tabs = React.forwardRef(function Tabs(inProps, ref) {
         resizeObserver.disconnect();
       }
     };
-  }, [updateIndicatorState, updateScrollButtonState]);
-
-  const handleTabsScroll = React.useMemo(
-    () =>
-      debounce(() => {
-        updateScrollButtonState();
-      }),
-    [updateScrollButtonState],
-  );
+  }, [updateIndicatorState]);
 
+  /**
+   * Toggle visibility of start and end scroll buttons
+   * Using IntersectionObserver on first and last Tabs.
+   */
   React.useEffect(() => {
+    let firstObserver;
+    let lastObserver;
+    const { 0: firstTab, length, [length - 1]: lastTab } = Array.from(tabListRef.current.children);
+    const threshold = 0.99;
+    const observerOptions = {
+      root: tabsRef.current,
+      threshold,
+    };
+
+    const handleScrollButtonStart = (entries) => {
+      let display = false;
+      entries.forEach(({ intersectionRatio }) => {
+        display = intersectionRatio < threshold;
+      });
+      setDisplayStartScroll(display);
+    };
+
+    const handleScrollButtonEnd = (entries) => {
+      let display = false;
+      entries.forEach(({ intersectionRatio }) => {
+        display = intersectionRatio < threshold;
+      });
+      setDisplayEndScroll(display);
+    };
+
+    if (
+      typeof IntersectionObserver !== 'undefined' &&
+      length > 0 &&
+      scrollable &&
+      scrollButtons !== false
+    ) {
+      firstObserver = new IntersectionObserver(handleScrollButtonStart, observerOptions);
+      firstObserver.observe(firstTab);
+
+      if (length > 1) {
+        lastObserver = new IntersectionObserver(handleScrollButtonEnd, observerOptions);
+        lastObserver.observe(lastTab);
+      }
+    }
+
     return () => {
-      handleTabsScroll.clear();
+      firstObserver?.disconnect();
+      lastObserver?.disconnect();
     };
-  }, [handleTabsScroll]);
+  }, [scrollable, scrollButtons, tabsRef, childrenProp]);
```

Comment:
Why is there `childrenProp` in the useEffect's dependency list? 

Also, refs (`tabsRef`) should not be in the dependency list because changing refs doesn't trigger a re-render for the effect to determine whether to run itself or not. Source - https://epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array/

Reply:
childrenProp we need it for the case of dynamic tabs (add or remove one).
So in dependency, do you prefer to do childrenProp.length or we keep it like that ?

Reply:
Yes, `childrenProp.length` is correct. `childrenProp` will always be a new reference on each and every render, so the effect may always run.

Reply:
Done :+1: